Substitute function options

janeclange opened the pull request:
Added -r flag for randomized subset selection
Added -b flag for breadth-first term and scope collection
Added "subset x of y" to testing logs for verbosity 2
Fixed a small typo?

janeclange commented:
added docstrings for _substitute, _filter_terms, and _filter_scopes 
removed pycache files

mpreiner commented on ddsmt.py:
```diff
@@ -128,13 +128,33 @@ def _test ():
     to_visit = [root if root else g_smtformula.scopes]
     while to_visit:
-        cur = to_visit.pop()
+        if bfs or g_args.bfs:
```
Can't we pass `g_args.bfs` to _filter_scopes (as option `bfs`) instead of checking `g_args.bfs` here?

mpreiner commented on ddsmt.py:
```diff
@@ -157,12 +177,30 @@ def _filter_cmds (filter_fun):
     visited = {}
     while to_visit:
-        cur = to_visit.pop().get_subst()
+        if bfs or g_args.bfs:
```
Ditto.

mpreiner commented on ddsmt.py:
```diff
@@ -184,10 +238,14 @@ def _substitute (subst_fun, substlist, superset, with_vars = False):
 
     while gran > 0:
-        subsets = [superset[s:s+gran] for s in range (0, len(superset), gran)]
+        if randomized or g_args.randomized:
```
Same here with `g_args.randomized`

janeclange commented:
passed g_args terms as parameters, changed docstrings to match

4tXJ7f commented:
This PR looks mostly good to me. This might be a good moment to discuss how we want to pass around options like `bfs`. I'd be in favor of either passing them as arguments through all the relevant functions and only reading them from `g_args` once or to always directly read them from `g_args` for consistency. Opinions?

mpreiner commented:
It depends on what you expect from the function. If you want to make it configurable via an argument `bfs` then only use that argument if you want to traverse in BFS order. If  you want the function to always use the option, don't provide a `bfs` argument. I would still read from `g_args` directly, but in the `bfs` case you might want to try both `false` and `true` after each other. Setting the option `g_args` for that internally is not a good solution.

4tXJ7f commented:
That makes sense. In that case, it seems like we'd want to also add `bfs` as an argument to `_filter_cmds` and `_substitute_terms` such that the main loop could vary that argument?

janeclange commented:
added bfs as a parameter for filter_cmds
added bfs and randomized as parameters for subst_terms, subst_cmds, and subst_scopes (the parameters don't really matter for scopes since it's by level anyway, but I added it for consistency)
more docstrings to reflect changes

janeclange commented:
Some small changes: 

- Previously,  _substitute would remove nodes from superset if they were not in any subset (only happens when randomized). I changed this so that nodes are only removed if they were in a successful test.

- Added a tracker for total time spent in the solver, which displays at verbosity 2.

mpreiner commented on ddsmt.py:
```diff
@@ -251,24 +340,38 @@ def _substitute_cmds (filter_fun = None):
+       command list cmds and are collected in the order indicated by the bfs parameter.
+
+       :subst_fun:  Function used to determine node substitutions.
+       :filter_fun: Function used to select terms to substitute.
```
Nit: Add additional space before `Function`.

mpreiner commented on ddsmt.py:
```diff
@@ -120,33 +120,66 @@ def _run (is_golden = False):
 
-def _filter_cmds (filter_fun):
+def _filter_cmds (filter_fun, bfs):
+    """ _filter_cmds(filter_fun, bfs)
```
Nit: Remove space between `"""` and `_filter_cmds`.

mpreiner commented on ddsmt.py:
```diff
@@ -157,12 +190,29 @@ def _filter_cmds (filter_fun):
+       :filter_fun: Boolean function that returns True if a node should be added.
+       :roots: List of nodes from which to begin searching.
+       :bfs: Bool indicating whether to use breadth-first search.
+       :return: List of term nodes that fit the filtering condition.
```
Ditto.

mpreiner commented on ddsmt.py:
```diff
@@ -120,33 +120,66 @@ def _run (is_golden = False):
+       :filter_fun: Boolean function that returns True if a node should be added.
+       :roots: List of nodes from which to begin searching.
+       :bfs: Bool indicating whether to use breadth-first search.
+       :return: List of scope nodes that fit the filtering condition.
```
Nit: Align text after parameters.

mpreiner commented on ddsmt.py:
```diff
@@ -120,33 +120,66 @@ def _run (is_golden = False):
+
+       :filter_fun: Boolean function that returns True if a node should be added.
+       :bfs: Bool indicating whether to use breadth-first search.
+       :return: List of command nodes that fit the filtering condition.
```
Ditto (align).
